README.md
IndexedRedis
============

A redis-backed very very fast ORM-style framework that supports indexes. It performs searches with O(1) efficency!

You can store and fetch native python types (lists, objects, strings, integers, etc.).

IndexedRedis supports both "equals" and "not-equals" operators for comparison. It also provides full atomic support for replacing entire datasets (based on model), which is useful for providing a fast frontend for SQL. In that use-case, a task that runs on an interval would fetch/calculate datasets from the SQL backend, and do an atomic replace on the datasets the front-end would query.


IndexedRedis since 6.0.0 also provides support for Foreign references ( like a foreign key one-to-many, many-to-one, or many-to-many relationship in SQL ) which allow you to directly convert your SQL models into Redis models for vastly improved performance.


Further client-side filtering (like greater-than, contains, etc) is available after the data has been fetched (see "Filter Results" below)

My tests have shown that for using equivalent models between flask/mysql and IndexedRedis, a 600% - 1200% performance increase occurs, yet if you design your storage directly as IndexedRedis models, you are able to achieve much higher gains.

It is compatible with python 2.7 and python 3. It has been tested with python 2.7, 3.4, 3.5, 3.6.


5.0 Series
----------

Version 5.0.0 will be somewhat **backwards incompatible with previous versions** by removing some old legacy stuff, improving a lot of existing things, and changing some behaviour.

You can get an overview of the steps needed to convert your code with the conversion guide, found at https://github.com/kata198/indexedredis/blob/5.0branch/CONVERTING_TO_5.0.0

Full details of changes, as well as enhancements and bug fixes omitted from the conversion guide can be found in the 5.0.0 Changelog: https://github.com/kata198/indexedredis/blob/5.0branch/Changelog


Automatic and Native Types
--------------------------

Since 4.0, IndexedRedis supports defining fields which will automatically be converted to/from native python types (such as int, float, datetime), as well as anything that can be represented with json (dicts, lists) or objects that support pickling. You just provide the type in its native format, and all the conversion happens behind the scenes. When fetched, the object returned also contains fields in their native types.

IndexedRedis also supports more advanced features such as automatically pickling/unpickling fields, compression/decompression, base64 encoding/decoding, and even defining your own custom field types through a standard interface.

See "Advanced Fields" section below for more information.


API Reference
-------------

Many, but not all methods and types are convered in this document.

For full pydoc reference, see:

https://pythonhosted.org/indexedredis/

or

http://htmlpreview.github.io/?https://github.com/kata198/IndexedRedis/blob/master/doc/IndexedRedis.html?_cache_vers=1


**Below is a quick highlight/overview:**


IndexedRedisModel
-----------------

This is the type you should extend to define your model.


**Example Model:**

	class Song(IndexedRedisModel):

		FIELDS = [ \
			IRField('artist'),
			IRField('title'),
			IRField('album'),
			IRField('track_number', valueType=int), # Convert automatically to/from int
			IRField('duration', defaultValue='0:00'),
			IRField('releaseDate', valueType=datetime.datetime),  # Convert automatically to/from datetime
			IRField('description'),
			IRField('copyright'),
			IRRawField('mp3_data'), # Don't try to encode/decode data
			IRCompressedField('thumbnail', compressMode='gzip'),      # Compress this field in storage using "gzip" compression
            IRField('tags', valueType=list),

            # "lyrics" will be a utf-8 unicode value on the object, and will be compressed/decompressed to/from storage
            IRFieldChain('lyrics', [ IRUnicodeField(encoding='utf-8'), IRCompressedField() ], defaultValue='No lyrics found' ),
		]

		INDEXED_FIELDS = [ \
					'artist',
					'title',
					'track_number',
		]

		KEY_NAME = 'Songs'


**Model Attributes:**


*FIELDS* - REQUIRED. A list of IRField objects (or their subclasses) which name the fields that can be used for storage. (see "Advanced Fields" section below)

	 Example: [IRField('name'), IRField('description'), IRField('model'), IRFixedPointField('Price', 2), IRField('timestamp', valueType=datetime), IRField('remainingStock', valueType=int)]


*INDEXED_FIELDS* - A list of strings containing the names of fields that will be indexed. Can only filter on indexed fields. Adds insert/delete time. The names listed here must match the name of a field given in FIELDS.

	 Example: ['Name', 'model']


*KEY_NAME* - REQUIRED. A unique name name that represents this model. Think of it like a table name.

	 Example: 'StoreItems'


*REDIS_CONNECTION_PARAMS* - OPTIONAL -  provides the arguments to pass into "redis.Redis", to construct a redis object. Here you can define overrides per-model from the default connection params.

Since 5.0.0, define this field ONLY for this model to use an alternate connection than the default. You no longer need to set this on every model.

See "Connecting To Redis" section below for more info.

If not defined or empty, the default params will be used. If any fields are present, they will override the inherited default connection params.

	 Example: {'host' : '192.168.1.1'}


Advanced Fields
---------------

IndexedRedis since version 4.0 allows you to pass elements of type IRField (extends str) in the FIELDS element.

Since 5.0.0, all fields must extend IRField in some way. Those that do not will generate a deprecated warning, and the field will be converted to an IRClassicField (same as IRField, but defaults to empty string instead of irNull).


Doing so allows you to specify certain properties about the field.


Example:

	FIELDS = [ IRField('name'), IRField('age', valueType=int), IRField('birthday', valueType=datetime.datetime) ]

**Field Name**

The first argument is the string of the field name.

**Type**

You can have a value automatically converted to a certain type on IRField (or use one of the several extending fields)

by passing that type as "valueType". (e.x.  IRField('age', valueType=int))

If you use "bool", the values 0 and case insensitive string 'false' will result in False, and 1 or 'true' will result in True.

When using floats, consider using IRFixedPointField, which supports indexing and the same representation regardless of platform (unlike "float"). 

floats to work cross-platform. Use a fixed point number as the string type ( like myFixedPoint = '%2.5f' %( 10.12345 ) )


IRField supports "valueType", most other field types deal with a specific type and thus don't have such a parameter.


**defaultValue**

All fields (except IRClassicField) support a parameter, given when constructing the IRField object, "defaultValue".

For all fields (except IRClassicField), the value of this parameter defaults to "irNull" (see below). For an IRClassicField, the default remains empty string and cannot be changed (to be compatible with plain-string fields pre-5.0.0).


**NULL Values**

Null values are represented by a static singleton, called "irNull" (of type IRNullType).

For all types except IRClassicField (which has a default of empty string) the default (when unset) value of the field is irNull. This can be changed by passing "defaultValue=somethingElse" to the IRField constructor.

irNull does not equal empty string, or anything except another irNull. This is to destinguish say, no int assigned vs int(0)

You can check a typed field against the "irNull" variable found in the IndexedRedis or IndexedRedis.fields.

For Example: 

	from IndexedRedis import irNull

..


*Null Values Can be used directly in the model filtering:*

	notDangerFive = MyModel.objects.filter(dangerLevel__ne=irNull).filter(dangerLevel__ne=5).all()

*or in results, through Queryable List, :*

	myResults = MyModel.objects.filter(something='value').all()

	notDangerFive = myResults.filter(dangerLevel__ne=irNull).filter(dangerLevel__ne=5)

*or direct comparison :*

	if someObj.specialFlagX == irNull and formFields.get('waiverForm') != 'checked':

		setError( 'We introduced X on 06/16/15. Customers who signed up prior may not have yet filled out form Y to assign them a category Z.\nIn order to use this feature, you must either fill out form Y in your properties section, call a service represenative, or perform a one-time waiver (checking the box next to the "waiver" signifies your signature of this waiver).


**Advanced Types**

The following are the possible field types, for use within the FIELDS array, and can be imported like "from IndexedRedis.fields import NAME":


**IRField** - Standard field, takes a name and a "valueType", which is a native python type, or any type you create which implements \_\_new\_\_, taking a signle argument and returning the object. See IndexedRedis/fields/FieldValueTypes for example of how datetime and json are implemented.

When no valueType is defined, str/unicode is the type (same as pre-4.0), and default encoding is used (see set/getDefaultIREncoding functions)

Indexable unless type is a json type or float (use IRFixedPointField to index on floats)


**IRBase64Field** - Converts to and from Base64.

Indexable.


**IRCompressedField** - Automatically compresses before storage and decompresses after retrieval. Argument "compressMode" currently supports "zlib" (default), "bz2", or "lzma".

Indexable.


**IRFixedPointField** - A floating-point with a fixed number of decimal places. This type supports indexing using floats, whereas IRField(...valueType=float) does not, as different platforms have different accuracies, roundings, etc. Takes a parameter, decimalPlaces (default 5), to define the precision after the decimal point.

Indexable.


**IRPickleField** - Automaticly pickles the given object before storage, and unpickles after fetch.

Not indexable because different representation between python2 and 3, and potentially system-dependent changes repr


**IRUnicodeField** - Field that takes a parameter, "encoding", to define an encoding to use for this field. Use this to support fields with arbitrary encodings, as IRField will use the default encoding for strings.

Indexable


**IRBytesField** - Field that forces the data to be "bytes", python2 and python3 compatible. If you need python3 only, you can use IRField(valueType=bytes). For no encoding/decoding at all, see IRRawField

Indexable


**IRClassicField** - Field that imitates the behaviour of a plain-string entry in FIELDS pre-5.0.0. This field has a default of empty string, and is always encoded/decoded using the defaultIREncoding

Indexable


**IRRawField** - Field that is not converted in any, to or from Redis. On fetch this will always be "bytes" type (or str in python2). On python3 this is very similar to IRField(...valueType=None), but python2 needs this to store binary data without running into encoding issues.

Not indexable - No decoding


**IRForeignLinkField** - Field that provides reference to a different model ( think "foreign key" in SQL ). Use this to reference other models from your model. This field links to a single foreign object, or irNull.

Takes the linked model as the "foreignKey" argument.

see "Foreign Links" section for more info.

Indexable


**IRForeignMultiLinkField** - Field that provides reference to a different model ( think "foreign key" in SQL) ). Use this to reference multiple other models from your model. This model links to one or more foreign objects, or irNull.

Takes the linked model as the "foreignKey" argument.

see "Foreign Links" section for more info.

Indexable. Note, filter must contain the full list (either pks, objs, or combination thereof). A "contains item"-style filter must be performed client-side.


**IRFieldChain** - Chains multiple field types together. Use this, for example, to compress the base64-representation of a value, or to compress utf-16 data. See section below for more details.

Indexable if all chained fields are indexable.


**Chaining Multiple Types**

"Chaining" allows you to apply multiple types on a single field. Say, for example, that you have some utf-16 data that you want to be compressed for storage:

Example:

	FIELDS = [ \

	...

		IRFieldChain( 'longData', [ IRUnicodeField(encoding='utf-16'), IRCompressedField() ] )
	]


An IRFieldChain works similar to a regular IRField, the first parameter is the field name, it has an optional "defaultValue" parameter.

The difference is that the second parameter, *chainedFields*, takes a list of other field types.

When storing, the value is passed through each type in this list, left-to-right.

When fetched, the value retrieved is passed backwards through these chainedFields, right-to-left.

The output of the leftmost (first) element is what defines the type of data that will be found on the object when accessed.

So in the above example, "myObj.longData" would be a utf-16 string. When going to the database, that utf-16 string will be decoded and then compressed for storage. When fetched, it will be decompressed and then converted back into utf-16.


You can specify a defaultValue on an IRFieldChain by providing "defaultValue=X" as an argument to the constructor. If you provide "defaultValue" on any of the fields in the chain list, however, it will be ignored.


**Hash-Lookups (performance)**


If you want to index/search on very large strings/bytes (such as maybe a genome), IndexedRedis supports hashing the key, i.e. the value will be stored as the value itself, but the key reference used for lookup will be a hash of that string.

This increases performance, saves network traffic, and shrinks storage requirements.


To do this, set the "hashIndex" attribute of an IRField to True.

	FIELDS = [ \

	...
		IRField ( 'genomeStr', hashIndex=True )
	]

and that's it! Filter and fetch and all operations remain the same (i.e. you just use the value directly, same as if "hashIndex" was False), but behind-the-scenes the lookups will all be done with the MD5 hash of the value.


**Converting existing models to/from hashed indexes**


IndexedRedis provides helper methods to automatically convert existing unhashed keys to hashed, and also hashed keys back to unhashed.

To do this, change your IndexedRedisModel accordingly, and then call (for a model class named MyModel):

	MyModel.objects.compat_convertHashedIndexes()

This will delete both the hashed and non-hashed key-value for any IRField which supports the "hashIndex" property.
If you just call "reindex" and you've changed the property "hashIndex" on any field, you'll be left with lingering key-values.

This function, by default (fetchAll=True) will fetch all records of this paticular model, and operate on them one-by-one. This is more efficient, but if memory constraints are an issue, you can pass fetchAll=False, which will fetch one object, convert indexes, save, then fetch next object. This is slower, but uses less memory.

NOTHING should be using the models while this function is being called (it doesn't make sense anyway to change schema whilst using it).


Connecting to Redis
-------------------

Your connection to Redis should be defined by calling "setDefaultRedisConnectionParams" with a dict of { 'host' : 'hostname', 'port' : 6379, 'db' : 0 }.

The default connection will connect to host at 127.0.0.1, port at 6379, and db at 0. If you don't define any of these fields explicitly, those values will be used for the respective field.


These default params will be used for all models, UNLESS you define REDIS\_CONNECTION\_PARAMS on a model to something non-empty, then that model will inherit the default connection parameters, overriding any values with those defined on the model.

If you need the same model to connect to different Redis instances, you can call "MyModel.connectAlt" (where MyModel is your model class) and pass a dict of alternate connection parameters. That function will return a copy of the class that will use the alternate provided connection.


Model Validation
----------------

The model will be validated the first time an object of that type is instantiated. If there is something invalid in how it is defined, an "InvalidModelException" will be raised.


Usage
-----

Usage is very similar to Django or Flask.

**Query:**


Changelog
6.0.4 - ??? ??? ?? ????

- Update runTests.py from 2.1.0 to 3.0.5. See
https://github.com/kata198/GoodTests for more details.

6.0.3 - Tue May 23 2017

- Try to make deepcopy, if possible, when setting/fetching values to _origData
(that is used to determine if a field has changed and needs to be updated on a
save). An example would be a stored json with a inside. If it was updated
externally, it would not be seen as an update. Now it will.

6.0.2 - Wed May 17 2017

- Update a bunch of docstrings
- Fix "valueType" missing on IRForeign*LinkField .copy and .getReprProperties
methods


6.0.1 - Wed May 17 2017

- Fix IRForeignLinkField not showing up in __all__ in fields.foreign, which
caused that field to not show up in pydoc


6.0.0 - Tue May 16 2017

Foreign Link Support! (Link/Fetch/Save objects related to other objects, like
foreign keys)

Also, if upgrading from <5.0.0, see CONVERTING_TO_5.0.0
and ChangeLog for possibly backwards-incompatible changes made in 5.0.0.

This release has no backwards incompatible changes.


- Add an IRForeignLinkField which allows linking another object by primary
key. You can assign it an instance of another model or a primary key, and upon
access if not already fetched the parent model will fetch the foreign object
and return it.

- Add IRForeignMultiLinkField which links one object to a list of other
objects.

^ Now that foreign fields are enabled, you can do a 1:1 conversion of your SQL
models into IndexedRedis and use them the same way. You CAN get better by
designing it for IR, but you WILL get about 5 to 15 times the performance at
least just by switching the engine.

- Add "cascadeSave" option to saving functions. This flag (default True) will
cause all foreign link objects to be saved (inserted or updated), recursively.

- Add "cascadeFetch" option to fetching functions. This flag (default False)
will cause all foreign link objects to be fetched right away, recursively. By
default, they are fetched on first-access.

- Add "cascadeObjects" to comparison functions. This controls whether the
values on the foreign objects are considered, or if just the pk relationship
is considered.

- Add "cascadeObjects" to reload function. This controls whether if any
foreign objects have changed values, we reload them, or only if the pk was
changed.

^ See "Foreign Links" section in README for details now.

- Some performance improvements

- Add "diff" method to IndexedRedisModel, to compare the values on two
objects, and return a dict of "changedField" : ( valueOnFirstObj,
valueOnSecondObj )


- Introduce "optional patches", as root-dir "patches" folder. These are
optional patches which may match your situation, but won't be included in
mainline for some reason.

There's a README in that dir which will list a brief explanation of
each optional patch, with a more descript in the comment field of the unified
diff in the same directory.


5.0.2 - Sat May 13 2017
- Fix where PyDoc could fail to create docs for projects importing
IndexedRedis in some circumstances
- Ensure that getPrimaryKeys function always returns a list and not sometimes
a set (because of upstream Redis package)


5.0.1 - Mon May 1 2017

A conversion document has been created, CONVERTING_TO_5.0.0 in the main
directory of this package. Also available online, at 
  https://github.com/kata198/indexedredis/blob/5.0branch/CONVERTING_TO_5.0.0

Use this document to prepare your code for the changes in IndexedRedis 5.0.0.
Also please review this Changelog to see new features, bug fixes, etc which
will not be listed in the conversion document.

- Improve IRPickleField to work with ALL types (including now, strings and
bytes)

  THIS WAS PREVIOUSLY BACKWARDS-INCOMPATIBLE, BUT HAS BEEN MADE COMPATIBLE
  WITHOUT CHANGES TO YOUR CODE. **DO NOT RUN THE COMPAT FUNCTION SHIPPED WITH
  4.1.3 - 4.1.4**


- Fixup the way connecting to Redis works. Now, instead of having to define
REDIS_CONNECTION_PARAMS on each model, you can call
"setDefaultRedisConnectionParams" to set a default connection configuration.
This configuration will be used on all models unless REDIS_CONNECTION_PARAMS has been
defined and is not empty.

- Rename "connect" method to "connectAlt" on IndexedRedisModel (to better
convey what it does). Also fix it up to better support multiple alternate
connection models

- Remove formerly deprecated IndexedRedisModel.BASE64_FIELDS - You've been
warned for a while now to use fields.IRField.IRBase64Field
  **BACKWARDS INCOMPATIBLE**
  To convert, remove the BASE64_FIELDS array from your model. For any items
  that are in it, change it in the FIELDS array to be:
    IRBase64Field
  as found in
    from IndexedRedis.fields import IRBase64Field
  
  If you already have a type, and were using it to base64 encode that type
  (like maybe you are compressing then base64 encoding), use a IRFieldChain
  with base64 as the last element to retain same functionality.

  Like:
     IRFieldChain('fieldName', [IRUnicodeField(encoding='utf-16'),
	 IRBase64Field()])

  Models with BASE64_FIELDS defined will generate an error upon validation.

- Remove IndexedRedis.BINARY_FIELDS - Use the new IRBytesField to act the same
way as a BINARY_FIELD did on python3 or python2, use IRField(valueType=bytes)
[python 3 only], or use IRRawField for no encoding/decoding at all. You can
combine these in a chain just like IRBase64Field.
  **BACKWARDS INCOMPATIBLE**

  Models with BINARY_FIELDS defined will generate an error upon validation.

- Make copy.copy and copy.deepcopy on an IndexedRedisModel call
IndexedRedisModel.copy(False), in order to copy the data, but NOT to have the
copy be linked to the database. A save on the copy will be an insert, the
parent will remain an update.
  **BACKWARDS INCOMPATIBLE**

- Removed the old, deprecated, compat global "defaultEncoding", retaining just
defaultIREncoding. Also, removed deprecated getEncoding and setEncoding, use getDefaultIREncoding and setDefaultIREncoding
  **BACKWARDS INCOMPATIBLE**


example.py
#!/usr/bin/env python
# vim: set ts=4 sw=4 st=4 expandtab :

import time

startTime = time.time()

import datetime
import sys
import IndexedRedis
from IndexedRedis import IndexedRedisModel, IRField, setDefaultRedisConnectionParams
from IndexedRedis.fields import IRCompressedField, IRFieldChain, IRRawField, IRBytesField, IRUnicodeField

# vim: set ts=8 sw=8 st=8 expandtab :

#  Change this to match your Redis server info
REDIS_CONNECTION_PARAMS = { 'host' : '127.0.0.1', 'port' : 6379, 'db' : 0 }

# Define the model
class Song(IndexedRedisModel):
    
    FIELDS = [ \
            IRField('artist'),
            IRField('title'),
            IRField('album'),
            IRField('track_number', valueType=int), # Convert automatically to/from int
            IRField('duration'),
            IRField('releaseDate', valueType=datetime.datetime),
            IRField('description'),
            IRField('copyright'),
            IRRawField('mp3_data'), # Do not perform any conversion on the data.
            IRFieldChain('thumbnail', [IRBytesField(), IRCompressedField(compressMode='gzip')]),      # Compress this field in storage using "gzip" compression
            IRField('tags', valueType=list),
            IRFieldChain('lyrics', [ IRUnicodeField(encoding='utf-8'), IRCompressedField() ], defaultValue='No lyrics found'),
    ]

    INDEXED_FIELDS = [ \
                'artist',
                'title',
                'track_number',
    ]

    KEY_NAME = 'Songs'

if __name__ == '__main__':

    setDefaultRedisConnectionParams(REDIS_CONNECTION_PARAMS)

    fakeMp3 = b"\x99\x12\x14"

    fakeThumbnail = b"\x15\x1A\x1A\x1A\x1A\x1A\x1A\x1B\x1A\x1A\x1A\x1B" + (b"\x1A" * 30) # Compressable data

    sys.stdout.write('Testing IndexedRedis version %s\n' %(IndexedRedis.__version__,))

    if '--keep-data' not in sys.argv:
            newSongs = []
            # Add data
            songObj = Song(artist='The Merry Men',
                            title='Happy Go Lucky',
                            album='The Merry Men LP',
                            track_number=1,
                            duration='1:58',
                            description='A song about happy people',
                            mp3_data=fakeMp3,
                            thumbnail=fakeThumbnail,
                            tags=['happy', 'guitar', 'smooth'],
                            copyright='Copyright 2012 (c) Media Mogul Incorporated')

            songObj.lyrics = '''To love, to lie,
to feel something rather than die
A line, a fall
I guess we can't be loved by them all..
'''

            newSongs.append(songObj)

            songObj = Song(artist='The Merry Men',
                            title='Joy to Joy',
                            album='The Merry Men LP',
                            track_number=2,
                            duration='2:54',
                            description='A song about joy',
                            mp3_data=fakeMp3,
                            thumbnail=fakeThumbnail,
                            tags=['joyful', 'piano', 'Key C'],
                            copyright='Copyright 2012 (c) Media Mogul Incorporated')

            newSongs.append(songObj)

            songObj = Song(artist='The Unhappy Folk',
                            title='Sadly she waits',
                            album='Misery loses comfort',
                            track_number=1,
                            duration='15:44',
                            description='A sad song',
                            mp3_data=fakeMp3,
                            thumbnail=fakeThumbnail,
                            copyright='Copyright 2014 (c) Cheese Industries')
            newSongs.append(songObj)

            # Atomically reset the Song dataset to just the songs in "newSongs" list
            Song.reset(newSongs)

            # Add some additional songs one-at-a-time
            songObj = Song(artist='Mega Men',
                            title='Nintendo 1',
                            album='Super Tracks',
                            track_number=1,
                            duration='1:15',
                            description='Super Nintendo',
                            mp3_data=fakeMp3,
                            thumbnail=fakeThumbnail,
                            copyright='Copyright 2014 (c) Cheese Industries')
            songObj.save()

            songObj = Song(artist='Mega Men',
                            title='Nintendo 2',
                            album='Super Tracks',
                            track_number=2,
                            duration='1:55',
                            description='Super Nintendo',
                            mp3_data=fakeMp3,
                            thumbnail=fakeThumbnail,
                            copyright='Copyright 2014 (c) Cheese Industries')
            songObj.save()


    # Query some songs by artist
    merryMenSongs = Song.objects.filter(artist='The Merry Men').all()

    sys.stdout.write("Merry Men Songs:\n")
    merryMenSongs.pprint()

    # Query some songs not by artist
    sys.stdout.write('\n\nNot Mega Men Songs:\n')
    notMegaMenSongs = Song.objects.filterInline(artist__ne='Mega Men').all()
    notMegaMenSongs.pprint()


    sys.stdout.write('\n\n')
    # Show passing filter objects around functions, and not actually fetching until .all is called.
    def getTracks(filterSet, trackNo):
            return filterSet.filter(track_number=trackNo).all()

    sys.stdout.write('\nAll track one songs:\n')

    allTrackOneSongs = getTracks(Song.objects, 1)
    for song in allTrackOneSongs:
            song.pprint()
            sys.stdout.write('\n')

    sys.stdout.write('\nMega Men track ones:')
    megaMenTracks = Song.objects.filter(artist='Mega Men')


    objs = getTracks(megaMenTracks, 1)
    for song in objs:
            song.pprint()
            sys.stdout.write('\n')

    sys.stdout.write('\nMega Men track twos (should be just one entry):\n')

    objs = getTracks(megaMenTracks, 2)
    for song in objs:
            song.pprint()
            sys.stdout.write('\n')

    pk = song.getPk()
    sys.stdout.write('\nPrimary key of previous song: %s\n' %(pk,))

    sys.stdout.write('\nSong pk=%s exists? (should be True) %s\n' %(pk, str(Song.objects.exists(pk)) ) )

    sys.stdout.write('\nAfter Delete, Mega Men Track twos (should be blank):\n')
    song.delete()
    objs = getTracks(megaMenTracks, 2)
    for song in objs:
            song.pprint()
            sys.stdout.write('\n')

    sys.stdout.write('\nSong pk=%s exists? (should now be False) %s\n' %(pk, str(Song.objects.exists(pk)) ) )
    
    # delete remaining Mega Men songs
    numDeleted = Song.objects.filter(artist='Mega Men').delete()
    sys.stdout.write('\nDeleting remaining Mega Men Tracks: %d deleted.\n' %(numDeleted,))

    sys.stdout.write('\nRemaining Mega Men tracks (should be blank):\n')
    songs = Song.objects.filter(artist='Mega Men').all()
    sys.stdout.write(str(songs) + '\n')

    endTime = time.time()

    sys.stderr.write("TIME: %.6f\n" %(endTime - startTime, ))
        

# vim: set ts=4 sw=4 st=4 expandtab :



###############################
##         OUTPUT:            #
###############################

#Testing IndexedRedis version 4.1.4
#Merry Men Songs:
#{'album': 'The Merry Men LP',
# 'artist': 'The Merry Men',
# 'copyright': 'Copyright 2012 (c) Media Mogul Incorporated',
# 'description': 'A song about joy',
# 'duration': '2:54',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': ['joyful', 'piano', 'Key C'],
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Joy to Joy',
# 'track_number': 2}
#
#{'album': 'The Merry Men LP',
# 'artist': 'The Merry Men',
# 'copyright': 'Copyright 2012 (c) Media Mogul Incorporated',
# 'description': 'A song about happy people',
# 'duration': '1:58',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': ['happy', 'guitar', 'smooth'],
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Happy Go Lucky',
# 'track_number': 1}
#
#
#
#Not Mega Men Songs:
#{'album': 'The Merry Men LP',
# 'artist': 'The Merry Men',
# 'copyright': 'Copyright 2012 (c) Media Mogul Incorporated',
# 'description': 'A song about joy',
# 'duration': '2:54',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': ['joyful', 'piano', 'Key C'],
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Joy to Joy',
# 'track_number': 2}
#
#{'album': 'Misery loses comfort',
# 'artist': 'The Unhappy Folk',
# 'copyright': 'Copyright 2014 (c) Cheese Industries',
# 'description': 'A sad song',
# 'duration': '15:44',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': IRNullType(),
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Sadly she waits',
# 'track_number': 1}
#
#{'album': 'The Merry Men LP',
# 'artist': 'The Merry Men',
# 'copyright': 'Copyright 2012 (c) Media Mogul Incorporated',
# 'description': 'A song about happy people',
# 'duration': '1:58',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': ['happy', 'guitar', 'smooth'],
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Happy Go Lucky',
# 'track_number': 1}
#
#
#
#
#All track one songs:
#{'album': 'Super Tracks',
# 'artist': 'Mega Men',
# 'copyright': 'Copyright 2014 (c) Cheese Industries',
# 'description': 'Super Nintendo',
# 'duration': '1:15',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': IRNullType(),
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Nintendo 1',
# 'track_number': 1}
#
#{'album': 'Misery loses comfort',
# 'artist': 'The Unhappy Folk',
# 'copyright': 'Copyright 2014 (c) Cheese Industries',
# 'description': 'A sad song',
# 'duration': '15:44',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': IRNullType(),
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Sadly she waits',
# 'track_number': 1}
#
#{'album': 'The Merry Men LP',
# 'artist': 'The Merry Men',
# 'copyright': 'Copyright 2012 (c) Media Mogul Incorporated',
# 'description': 'A song about happy people',
# 'duration': '1:58',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': ['happy', 'guitar', 'smooth'],
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Happy Go Lucky',
# 'track_number': 1}
#
#
#Mega Men track ones:{'album': 'Super Tracks',
# 'artist': 'Mega Men',
# 'copyright': 'Copyright 2014 (c) Cheese Industries',
# 'description': 'Super Nintendo',
# 'duration': '1:15',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': IRNullType(),
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Nintendo 1',
# 'track_number': 1}
#
#
#Mega Men track twos (should be just one entry):
#{'album': 'Super Tracks',
# 'artist': 'Mega Men',
# 'copyright': 'Copyright 2014 (c) Cheese Industries',
# 'description': 'Super Nintendo',
# 'duration': '1:55',
# 'mp3_data': b'\x99\x12\x14',
# 'releaseDate': IRNullType(),
# 'tags': IRNullType(),
# 'thumbnail': b'\x15\x1a\x1a\x1a\x1a\x1a\x1a\x1b\x1a\x1a\x1a\x1b'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a\x1a'
#              b'\x1a\x1a\x1a\x1a\x1a\x1a',
# 'title': 'Nintendo 2',
# 'track_number': 2}
#
#
#Primary key of previous song: 6
#
#Song pk=6 exists? (should be True) True
#
#After Delete, Mega Men Track twos (should be blank):
#
#Song pk=6 exists? (should now be False) False
#
#Deleting remaining Mega Men Tracks: 1 deleted.
#
#Remaining Mega Men tracks (should be blank):
#IRQueryableList([])


# vim: set ts=4 sw=4 st=4 expandtab :
